src/test/unit/io/deserializer_test.cpp

```diff
@@ -0,0 +1,899 @@
+  EXPECT_FLOAT_EQ(0.0 + 1.0 * inv_logit_m2,
+                  (deserializer.read_lub<double, true>(0.0, 1.0, lp)));
+  EXPECT_FLOAT_EQ(3.0 + 2.0 * inv_logit_3,
+                  (deserializer.read_lub<double, true>(3.0, 5.0, lp)));
+  EXPECT_FLOAT_EQ(-3.0 + 5.0 * inv_logit_m1,
+                  (deserializer.read_lub<double, true>(-3.0, 2.0, lp)));
+  EXPECT_FLOAT_EQ(-15.0 + 30.0 * inv_logit_0,
+                  (deserializer.read_lub<double, true>(-15.0, 15.0, lp)));
+  double expected_lp = -7.2
+                       + log((1.0 - 0.0) * inv_logit_m2 * (1 - inv_logit_m2))
+                       + log((5.0 - 3.0) * inv_logit_3 * (1 - inv_logit_3))
+                       + log((2.0 - -3.0) * inv_logit_m1 * (1 - inv_logit_m1))
+                       + log((15.0 - -15.0) * inv_logit_0 * (1 - inv_logit_0));
+  EXPECT_FLOAT_EQ(expected_lp, lp);
+}
+
+// offset multiplier
+TEST(deserializer_scalar, offset_multiplier_constrain) {
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  theta.push_back(-2.0);
+  theta.push_back(3.0);
+  theta.push_back(-1.0);
+  theta.push_back(0.0);
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+  double lp = 0;
+  EXPECT_FLOAT_EQ(
+      -2.0, (deserializer.read_offset_multiplier<double, false>(0.0, 1.0, lp)));
+  EXPECT_FLOAT_EQ(
+      3.0 + 5.0 * 3.0,
+      (deserializer.read_offset_multiplier<double, false>(3.0, 5.0, lp)));
+  EXPECT_FLOAT_EQ(
+      -3.0 + 2.0 * -1.0,
+      (deserializer.read_offset_multiplier<double, false>(-3.0, 2.0, lp)));
+  EXPECT_FLOAT_EQ(-15.0, (deserializer.read_offset_multiplier<double, false>(
+                             -15.0, 15.0, lp)));
+}
+
+TEST(deserializer_scalar, offset_multiplier_constrain_jacobian) {
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  theta.push_back(-2.0);
+  theta.push_back(3.0);
+  theta.push_back(-1.0);
+  theta.push_back(0.0);
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+  double lp = -7.2;
+  EXPECT_FLOAT_EQ(
+      -2.0, (deserializer.read_offset_multiplier<double, true>(0.0, 1.0, lp)));
+  EXPECT_FLOAT_EQ(
+      3.0 + 5.0 * 3.0,
+      (deserializer.read_offset_multiplier<double, true>(3.0, 5.0, lp)));
+  EXPECT_FLOAT_EQ(
+      -3.0 + 2.0 * -1.0,
+      (deserializer.read_offset_multiplier<double, true>(-3.0, 2.0, lp)));
+  EXPECT_FLOAT_EQ(-15.0, (deserializer.read_offset_multiplier<double, true>(
+                             -15.0, 15.0, lp)));
+  double expected_lp = -7.2 + log(1.0) + log(5.0) + log(2.0) + log(15.0);
+  EXPECT_FLOAT_EQ(expected_lp, lp);
+}
+
+// unit vector
+
+TEST(deserializer_vector, unit_vector_constrain) {
+  std::vector<int> theta_i;
+  std::vector<double> theta;
+  theta.push_back(3.0);
+  theta.push_back(-1.0);
+  theta.push_back(-2.0);
+  theta.push_back(0.0);
+  stan::io::deserializer<double> deserializer(theta, theta_i);
+  double lp = 0;
+  Eigen::VectorXd reference
+      = stan::math::unit_vector_constrain(stan::math::to_vector(theta));
```

Comment:
I changed a lot of the tests to just check against Math library functions. Like at this point we're just testing the deserializer interface and we're assuming these functions are good, so I think it's fair. Makes it a lot easier to code and check jacobians and stuff.